Add option to install KW systemwide

uwla opened the pull request:
Add `--system` flag in `setup.sh` for systemwide installation.

This enhancement brings the following benefits:

1. System wide installation allows several users to run kw.
2. System wide installation does not require modifying PATH, such as in .bashrc.
3. It can be useful when testing because some testing environments (such as docker or podman containers) do NOT have $HOME/.local/bin in their $PATH and do NOT read .bashrc files by default, which requires manually changing PATH.

Closes #985

rodrigosiqueira commented on setup.sh:
```diff
@@ -18,16 +19,16 @@ declare -r app_name='kw'
-declare -r mandir="$sharedir/man"
-declare -r sounddir="$sharedir/sound"
-declare -r databasedir="$sharedir/database"
+declare kwbinpath="$HOME/.local/bin/$app_name"
+declare binpath="$HOME/.local/bin"
+declare libdir="$HOME/.local/lib/$app_name"
```
https://kworkflow.org/content/codingstyle.html#string-concatenation

rodrigosiqueira commented on kw:
```diff
@@ -22,9 +22,10 @@ KW_ETC_DIR="${XDG_CONFIG_HOME:-"${HOME}/.config"}/${KWORKFLOW}"
   KW_SOUND_DIR='/usr/share/sounds/kw'
   KW_DOC_DIR='/usr/share/doc/kw/html/'
-  KW_SOUND_DIR='/usr/share/sounds/kw'
+  KW_MAN_DIR='/usr/share/man/en/kw'
```
Are you sure this path is correct?

uwla replied:
More or less sure. Debian keeps man pages at `/usr/share/` under their language abbreviation, such as: jp, ch, es, pt. That is why I put it in `/usr/share/man/en/kw` instead of `/usr/share/man/kw`.

rodrigosiqueira replied:
Tbh, I'm a little bit confused about the man path.

I checked this path in my Debian and Arch Linux systems, and none have the subfolder `en` (my system is set to English). I'm under the impression that the default English documentation for a command publicly available for the system should be in the subfolder `man1`. However, I'm not 100% sure... let's try to be sure about this.

rodrigosiqueira commented on kw:
```diff
@@ -22,9 +22,10 @@ KW_ETC_DIR="${XDG_CONFIG_HOME:-"${HOME}/.config"}/${KWORKFLOW}"
 # Set system-wide path
 if [[ -f "$KW_SYSTEM_WIDE_INSTALLATION" ]]; then
   KW_LIB_DIR='/usr/share/kw'
+  KW_DB_DIR="/usr/share/kw/database"
```
We don't want kw database in a protected directory; it is better to be at the user's home.

uwla replied:
Ok.

rodrigosiqueira commented on setup.sh:
```diff
@@ -18,16 +19,16 @@ declare -r app_name='kw'
-declare -r mandir="$sharedir/man"
-declare -r sounddir="$sharedir/sound"
-declare -r databasedir="$sharedir/database"
+declare kwbinpath="$HOME/.local/bin/$app_name"
```
Why remove `-r`? It looks unsafe...

uwla replied:
Because I need to change the value of the variable if it is system-wide installation. It will be at `/usr/bin/kw`. It can't be readonly, therefore.

rodrigosiqueira commented on setup.sh:
```diff
@@ -226,6 +232,7 @@ function usage()
   say "--uninstall | -u     Uninstall $app_name"
   say '--skip-checks        Skip checks (use this when packaging)'
   say '--skip-docs          Skip creation of man pages (use this when installing)'
+  say '--system    | -s     Install kw system wide (use this when installing)'
```
You also need to manage the uninstall in this scenario.

uwla replied:
Yes, we managed uninstall too. I left some comments about it in the code:

```bash
# if system wide, change the installation directories.
# This approach works both for installing and uninstalling KW.
if [[ "${SYSTEMWIDE}" == 1 ]]; then
  use_system_wide_installation_directories
fi
```

uwla commented on kw:
```diff
@@ -22,9 +22,10 @@ KW_ETC_DIR="${XDG_CONFIG_HOME:-"${HOME}/.config"}/${KWORKFLOW}"
+  KW_DB_DIR="/usr/share/kw/database"
   KW_SOUND_DIR='/usr/share/sounds/kw'
   KW_DOC_DIR='/usr/share/doc/kw/html/'
-  KW_SOUND_DIR='/usr/share/sounds/kw'
```
Look that there is already `KW_SOUND_DIR` in line 26. Therefore, line 27 was unnecessary duplicated code.

uwla commented:
@rodrigosiqueira I have addressed the comments and here is how we can test the PR:

Install kw in a podman container to test system-wide installation:

```
$ kw_path=/home/user/code/kworkflow # change it accordingly
$ podman run -v "${kw_path}":/kw -e PATH='/usr/bin' -w /kw --rm -it docker.io/library/archlinux bash
# ./setup.sh --install --system --force
```

The `$` prefix denotes running shell in host machine as a normal user.
The `#` prefix denotes running shell in container machine as root user.

Check the files were installed:

```
# find /usr ~/ -type f | grep /kw
...
```

Test commands:

```
# kw --version
# kw config --show mail
# kw patchhub
# kw mail --help
```

Uninstall:

```
# ./setup.sh --completely-remove --force --system
```

Check the files were uninstalled:

```
# find /usr ~/ -type f | grep /kw
```


By the way: the above commands are kind of a sketch for a `install_kw_system_wide` integration test.

rodrigosiqueira commented on setup.sh:
```diff
@@ -18,16 +19,16 @@ declare -r app_name='kw'
-declare -r mandir="$sharedir/man"
-declare -r sounddir="$sharedir/sound"
-declare -r databasedir="$sharedir/database"
+declare kwbinpath="${HOME}/.local/bin/$app_name"
+declare binpath="${HOME}/.local/bin"
+declare libdir="${HOME}/.local/lib/$app_name"
+declare sharedir="${XDG_DATA_HOME:-"$HOME/.local/share"}/$app_name"
+declare docdir="${sharedir}/doc"
+declare mandir="${sharedir}/man"
+declare sounddir="${sharedir}/sound"
+declare databasedir="${sharedir}/database"
```
Consider making another commit just to use the upper case for all of these constant variables since it is more aligned with the overall code style.

uwla replied:
ok

rodrigosiqueira commented on setup.sh:
```diff
@@ -166,6 +167,11 @@ function update_path()
   local shellrc=${1:-'.bashrc'}
 
+  # there is no need to update the path if the installation is system wide
+  if [[ "${SYSTEMWIDE}" == 1 ]]; then
```
I think you can just use `[[ -n "${SYSTEMWIDE}" ]]`.

rodrigosiqueira commented on setup.sh:
```diff
@@ -10,6 +10,7 @@ VERBOSE=0
 FORCE=0
 SKIPCHECKS=0
 SKIPDOCS=0
+SYSTEMWIDE=0
```
How about `SYSTEM_WIDE=''

rodrigosiqueira commented on setup.sh:
```diff
@@ -445,7 +453,11 @@ function synchronize_files()
   fi
 
-  say "$app_name installed into $HOME"
+  if [[ "$SYSTEMWIDE" == 1 ]]; then
```
How about `[[ -n "$SYSTEM_WIDE" ]]`.

rodrigosiqueira commented on setup.sh:
```diff
@@ -502,15 +514,36 @@ Commit: $head_hash
+  # The exception is the ETC directory and the BIN path. System wide config
+  # files are usually located at /etc and not at /usr while binary files are
+  # usually at /usr/bin
+  binpath="/usr/bin"
```
single quotes.

rodrigosiqueira commented on setup.sh:
```diff
@@ -502,15 +514,36 @@ Commit: $head_hash
   fi
-  # Move old folder structure to new one
-  legacy_folders
+  if [[ "$SYSTEMWIDE" == 0 ]]; then
```
Same as my previous comments.

rodrigosiqueira commented on src/help.sh:
```diff
@@ -54,13 +54,13 @@ function kworkflow_man()
 
-  if [[ -f "$KW_SYSTEM_WIDE_INSTALLATION" ]]; then
-    cmd_manager "$flag" "man ${feature}"
+  if [[ -r "$doc/$feature.1" ]]; then
+    cmd_manager "$flag" "man -l $doc/$feature.1"
```
String concatenation.

uwla commented:
@rodrigosiqueira , I have done some research, and I learned the following:

- Man searches for pages located in a directory defined by `MANPATH`. It usually searchs in `/usr/share/man` and `/usr/local/share/man`.
- Man pages are divided into sections (usually, from 1 to 9). If the man pages for a program has multiple sections, they would be at `${MANPATH}/man1/<programname>.1`, `${MANPATH}/man2/<programname>.2`, and so on. So, there is a suffix for the section.
- If the man page is not in English, it will be in a subdirectory denoted by the language acronym

uwla commented:
I closed it by mistake and I just realized it now....

uwla commented:
@rodrigosiqueira if you agree in putting kworkflow's manpages at /usr/share/man/man1/ then I can improve the PR very quickly.

uwla commented:
I pushed a commit but forgot to address one change. I'll do it later today.

uwla commented:
@rodrigosiqueira I think we have it now

uwla commented:
@rodrigosiqueira, when you have time, let's finish it.

rodrigosiqueira commented:
Hi @uwla ,
I'm sorry I never got to this important change. I'll revive it soon, and I'll keep your credits for future changes.
